volumes/miovision/sql/function-aggregate-volumes_15min_tmc.sql

```diff
@@ -10,62 +10,61 @@ AS $BODY$
 
 BEGIN
 
+WITH zero_padding_movements AS (
+		/*Cross product of legal movement for cars, bikes, and peds and the bins to aggregate*/
+		SELECT m.*, datetime_bin15 
+		FROM miovision_api.intersection_movements_new m
+		CROSS JOIN generate_series(start_date - interval '1 hour', end_date - interval '1 hour 15 minutes', INTERVAL '15 minutes') AS dt(datetime_bin15)
+		WHERE classification_uid IN (1,2,6) 
+		)
+, aggregate_insert AS (
 INSERT INTO miovision_api.volumes_15min_tmc(intersection_uid, datetime_bin, classification_uid, leg, movement_uid, volume)
 
 SELECT 
-m.intersection_uid,
-dt.datetime_bin,
-m.classification_uid,
-m.leg,
-m.movement_uid,
+pad.intersection_uid,
+pad.datetime_bin15,
+pad.classification_uid,
+pad.leg,
+pad.movement_uid,
 CASE WHEN un.accept = FALSE THEN NULL ELSE (COALESCE(SUM(A.volume), 0)) END AS volume
-FROM miovision_api.intersection_movements_new m
---Cross product of legal movement for cars, bikes, and peds and the bins to aggregate
-CROSS JOIN generate_series(start_date - interval '1 hour', end_date - interval '1 hour 15 minutes', INTERVAL '15 minutes') AS dt(datetime_bin)
---To get 1min bins
-LEFT JOIN miovision_api.volumes A
-	ON A.intersection_uid = m.intersection_uid AND A.classification_uid = m.classification_uid
-	AND A.leg = m.leg AND A.movement_uid = m.movement_uid
-	AND dt.datetime_bin = datetime_bin_15(A.datetime_bin)
---To set unacceptable ones to NULL instead
+FROM zero_padding_movements pad
+--To set unacceptable ones to NULL instead (& only gap fill  light vehicles, cyclist and pedestrian)
 LEFT JOIN miovision_api.unacceptable_gaps un
-	ON un.intersection_uid = A.intersection_uid
-	AND DATE_TRUNC('hour', un.gap_start) = DATE_TRUNC('hour', dt.datetime_bin) 	
---Only interested in light vehicles, cyclist and pedestrian
-WHERE m.classification_uid IN (1,2,6) 
+	ON un.intersection_uid = pad.intersection_uid
+	AND DATE_TRUNC('hour', un.gap_start) = DATE_TRUNC('hour', pad.datetime_bin15) 	
+--To get 1min bins
+FULL OUTER JOIN miovision_api.volumes A
+	ON datetime_bin >= start_date - INTERVAL '1 hour' 
+	AND datetime_bin < end_date - INTERVAL '1 hour'
+	AND A.intersection_uid = pad.intersection_uid 
+	AND A.classification_uid = pad.classification_uid
+	AND A.leg = pad.leg 
+	AND A.movement_uid = pad.movement_uid
+	AND A.datetime_bin >= pad.datetime_bin15 
+	AND A.datetime_bin < pad.datetime_bin15 + interval '15 minutes'
 -- make sure that the intersection is still active
-AND m.intersection_uid IN (SELECT intersection_uid FROM miovision_api.intersections_new 
+WHERE pad.intersection_uid IN (SELECT intersection_uid FROM miovision_api.intersections_new 
 				WHERE start_date::date > date_installed 
 				AND date_decommissioned IS NULL)
 AND A.volume_15min_tmc_uid IS NULL
-GROUP BY m.intersection_uid, dt.datetime_bin, m.classification_uid,m.leg,m.movement_uid, un.accept;
+GROUP BY pad.intersection_uid, pad.datetime_bin15, pad.classification_uid,pad.leg,pad.movement_uid, un.accept
```

Comment:
add `HAVING classification_uid IN (1,2,6) OR SUM(A.volume) > 0` so that rows with other classifications aren't included with `NULL`s